### PR TITLE
Replace ANTLR with CoreNLP in text frontend

### PR DESCRIPTION
**Pull request by SirYwell**
Targets #556.

With this PR, text is now tokenized by CoreNLP.

This has some more or less obvious effects:
- CoreNLP does not provide per-token line numbers. Therefore, we need to go through the file contents ourselves to keep track of line numbers and line break indexes to calculate columns. The behavior of supporting CR, LF, and CRLF stays.
- Without further adjustments, CoreNLP recognizes almost all non-whitespace symbols as tokens. Also, the splitting behavior differs (e.g. `1.5` was tokenized as `|1|5` before, now it's just `|1.5`). As a very simple way to get a somewhat similar behavior, I added the `isWord` method that checks if the token contains any alphanumeric symbols.
- CoreNLP also seems to recognize XML. In the javadoc test, it therefore creates one token for each tag.

There are likely more differences on other inputs.

A diff for the current output of the javadoc test can be found here: <https://gist.github.com/SirYwell/93c22c41af2adfdd549e56564c3260f2/revisions> (I simply pasted the original and edited with the current state)

I didn't change the existing test yet as I first would like to get some feedback on the current implementation.

(cc @tsaglam, I started working on this as discussed)

### CODE REVIEW

**Comment by tsaglam** [comment]
@SirYwell just FYI I am out of the office until 5.8, I will have a look at your work after that.

File: README.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ In the following, a list of all supported languages with their supported languag
 | [Scheme](http://www.scheme-reports.org)                          |       ? | scheme                | unknown | JavaCC |
 | [EMF Metamodel](https://www.eclipse.org/modeling/emf/)           |  2.25.0 | emf-metamodel         | alpha | EMF |
 | [EMF Metamodel](https://www.eclipse.org/modeling/emf/) (dynamic) |  2.25.0 | emf-metamodel-dynamic | alpha | EMF |
-| Text (naive)                                                     |       - | text                  | legacy | ANTLR |
+| Text (naive)                                                     |       - | text                  | legacy | CoreNLP |
```
Comment (JanWittler) [inline README.md:32]: I'd say we can remove the naive part here with this improvement

File: jplag.frontend.text/src/main/java/de/jplag/text/TokenPosition.java [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package de.jplag.text;
+
+public record TokenPosition(
+    int line,
+    int column,
+    int length
+) {
+}
```
Comment (JanWittler) [inline jplag.frontend.text/src/main/java/de/jplag/text/TokenPosition.java:8]: Please provide some minimal documentation here

File: jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java [hunk 3]
```diff
@@ -1,23 +1,36 @@
...
+import java.nio.file.Path;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
 
 public class ParserAdapter extends AbstractParser {
 
     private final Map<String, Integer> tokenTypes = new HashMap<>();
+    private final StanfordCoreNLP pipeline;
     private int tokenTypeIndex = 1; // 0 is FILE_END token, SEPARATOR is not used as there are no methods.
```
Comment (JanWittler) [inline jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java:22]: Seems there is an old bug in the text frontend. Please change this value to 2. Even though `SEPARATOR` is not used, as the constants don't have any identity, the value `1` will still be interpreted as it. Thus, we must not use it for custom token constants.

File: jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java [hunk 4]
```diff
@@ -1,23 +1,36 @@
...
 
     private TokenList tokens;
     private String currentFile;
+    private int currentLine;
+    private int currentLineIndex;
```
Comment (JanWittler) [inline jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java:27]: This is rather the rather the `globalCharIndex`, right? At least from the name it is not clear to me what the purpose of the attribute is.

File: jplag.frontend.text/src/main/java/de/jplag/text/TokenPosition.java (hunk 2 above)
Comment (JanWittler) [reply jplag.frontend.text/src/main/java/de/jplag/text/TokenPosition.java:8]: Or actually, you could also just remove this record, adjust the constructor of `TextToken` to take these 3 values and pass them directly. As far as I see, the record has no significance.

**Review by JanWittler — COMMENTED** [review]
Thanks for the improvement. The code looks overall good to me, I added some minor remarks.

As a remark for @jplag/maintainer: This PR increases the binary to ~70MB (see #586). We should keep an eye on this.

File: jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java (hunk 4 above)
Comment (SirYwell) [reply jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java:27]: It's the global index of the current line break. I'm not sure about a better name, but I'll add a comment explaining it. If you have a better name, please let me know.

File: README.md (hunk 1 above)
Comment (SirYwell) [reply README.md:32]: I don't have a strong opinion on this, but as we don't use any other CoreNLP feature than tokenzing, it's still somewhat naive.

File: jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java [hunk 5]
```diff
@@ -32,42 +49,75 @@ public TokenList parse(File directory, String[] files) {
...
-            inputState = new InputState(inputStream);
-            TextLexer lexer = new TextLexer(inputState);
-            lexer.setFilename(file);
-            lexer.setTokenObjectClass("de.jplag.text.AntlrParserToken");
+    private boolean isWord(CoreLabel token) {
+        // consider a token as a word if it contains any alphanumeric character
+        String text = token.originalText();
+        for (int i = 0; i < text.length(); i++) {
+            char c = text.charAt(i);
+            if (Character.isAlphabetic(c) || Character.isDigit(c)) {
```
Comment (tsaglam) [inline jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java:99]: You can use a foreach loop with `String.toCharArray()`. Makes it a bit more cleaner.

Comment (tsaglam) [reply jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java:99]: Or you can use a more functional one liner:
`return text.chars().anyMatch(it -> Character.isAlphabetic(it) || Character.isDigit(it))`

**Review by tsaglam — CHANGES_REQUESTED** [review]
Just a few minor comments.

File: jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java (hunk 4 above)
Comment (JanWittler) [reply jplag.frontend.text/src/main/java/de/jplag/text/ParserAdapter.java:31]: Then I would name it `currentLineBreakIndex` to make things clear.

File: README.md (hunk 1 above)
Comment (dfuchss) [reply README.md:32]: I would suggest to keep "naive" until some mechanism like lemma or stemming is added

**Review by tsaglam — APPROVED** [review]
Looks good!

File: README.md (hunk 1 above)
Comment (tsaglam) [reply README.md:32]: I agree with @dfuchss

**Comment by SirYwell** [comment]
If you're happy with the resulting tokens, I'd update the old test with the new numbers accordingly.

**Comment by tsaglam** [comment]
> If you're happy with the resulting tokens, I'd update the old test with the new numbers accordingly.

@SirYwell Yes, I think the resulting tokens look goo.! You can move ahead!

**Pull request merged**